commit b3e47525357322433126389533501c42cd092979
Author: Benda Xu <heroxbd@gentoo.org>
Date:   Sat Jun 18 14:32:07 2016 +0900

    do not use host PATH by default and prepend EPREFIX PATH.
    
      EPREFIX could be overridden in cross-eprefix, in that case tools
      inside EPREFIX should be prioritized.

diff --git a/pym/portage/package/ebuild/doebuild.py b/pym/portage/package/ebuild/doebuild.py
index 52dbf8b..304b164 100644
--- a/pym/portage/package/ebuild/doebuild.py
+++ b/pym/portage/package/ebuild/doebuild.py
@@ -199,16 +199,18 @@ def _doebuild_path(settings, eapi=None):
 	if portage_bin_path[0] != portage.const.PORTAGE_BIN_PATH:
 		# Add a fallback path for restarting failed builds (bug 547086)
 		portage_bin_path.append(portage.const.PORTAGE_BIN_PATH)
-	eprefix = portage.const.EPREFIX
 	prerootpath = [x for x in settings.get("PREROOTPATH", "").split(":") if x]
 	rootpath = [x for x in settings.get("ROOTPATH", "").split(":") if x]
 	overrides = [x for x in settings.get(
 		"__PORTAGE_TEST_PATH_OVERRIDE", "").split(":") if x]
 
 	prefixes = []
-	if eprefix:
-		prefixes.append(eprefix)
-	prefixes.append("/")
+	# tools in EPREFIX can only be executed when ROOT is /.
+	if settings["ROOT"] == "/":
+		prefixes.append(settings["EPREFIX"])
+	# settings["EPREFIX"] could be overridden during cross-eprefix
+	if portage.const.EPREFIX != settings["EPREFIX"]:
+		prefixes.append(portage.const.EPREFIX)
 
 	path = overrides
 
@@ -232,6 +234,7 @@ def _doebuild_path(settings, eapi=None):
 	path.extend(prerootpath)
 
 	for prefix in prefixes:
+		prefix = prefix if prefix else "/"
 		for x in ("usr/local/sbin", "usr/local/bin", "usr/sbin", "usr/bin", "sbin", "bin"):
 			path.append(os.path.join(prefix, x))
 
